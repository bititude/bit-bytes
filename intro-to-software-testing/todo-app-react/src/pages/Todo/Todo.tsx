import { useState } from 'react';
import AddTodo from './components/AddTodo/AddTodo';
import TodoLists from './components/TodoLists/TodoList';
import { useTodo } from './hooks/useTodo';

const Todo: React.FC = () => {
  const [isFormVisible, setIsFormVisible] = useState(false);
  const { todos, isTodosLoading, error } = useTodo();
  return (
    <div className='p-5 h-auto w-11/12 md:w-[450px]  shadow-md rounded-md m-auto border-2 border-gray-400 flex flex-col items-center justify-start mt-[100px] mb-20'>
      <div className='flex justify-between items-center mb-7 w-full'>
        <h1 className='font-bold text-gray-700 text-2xl'>Your todos</h1>
        <button
          data-testid='toggle-form-btn'
          className={`rounded-full ${
            isFormVisible ? 'bg-red-500' : 'bg-green-500'
          }  w-12 h-12 flex items-start justify-center hover:scale-105 shadow-md hover:shadow-lg`}
          onClick={() => setIsFormVisible(!isFormVisible)}
        >
          <span className='text-4xl text-white font-bold '>
            {isFormVisible ? <>&#xd7;</> : <>&#43;</>}
          </span>
        </button>
      </div>
      {isFormVisible ? <AddTodo /> : null}
      {error ? <p>{error.error}</p> : <TodoLists todos={todos} loading={isTodosLoading} />}
    </div>
  );
};

export default Todo;
