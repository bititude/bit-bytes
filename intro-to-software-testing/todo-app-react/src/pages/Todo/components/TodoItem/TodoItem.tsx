import { Todo } from '../../../../types/Todo';
import { useTodo } from '../../hooks/useTodo';

interface ITodoItem {
  todo: Todo;
}

const TodoItem: React.FC<ITodoItem> = ({ todo }) => {
  const { deleteTodo, markAsCompleted } = useTodo();

  return (
    <div
      className='border-2 border-gray-400 py-2 px-4 font-bold rounded-md shadow-md mb-4 flex items-center justify-between w-full'
      data-testid='todo'
    >
      {todo.completed ? (
        <p className='w-4/5 text-gray-500' data-testid='striked-todo'>
          {' '}
          <s className='decoration-2'>
            <i>{todo.title}</i>
          </s>
        </p>
      ) : (
        <p className='w-4/5'>{todo.title}</p>
      )}

      <button
        className='rounded-full bg-red-500   w-6 h-6 flex items-start justify-center hover:scale-105 '
        data-testid='todo-delete-btn'
        onClick={() => deleteTodo(todo.id)}
      >
        <span className=' text-white font-extrabold '>{<>&#xd7;</>}</span>
      </button>
      {!todo.completed ? (
        <button
          className='rounded-full bg-green-500 ml-3  w-6 h-6 flex items-start justify-center hover:scale-105 '
          data-testid='todo-markascomplete-btn'
          onClick={() => {
            markAsCompleted(todo.id);
          }}
        >
          <span className=' text-white font-extrabold '>{<>&#10003;</>}</span>
        </button>
      ) : null}
    </div>
  );
};

export default TodoItem;
