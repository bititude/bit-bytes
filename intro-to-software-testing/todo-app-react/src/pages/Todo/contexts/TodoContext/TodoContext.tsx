import { createContext, ReactNode, useCallback, useEffect, useState } from 'react';
import { Todo } from '../../../../types/Todo';
import {
  createTodo,
  getTodos,
  markTodoAsComplete,
  removeTodo,
} from '../../../../services/api/todo/todo';

type TodoError = {
  type: 'ADD' | 'LOAD' | 'DELETE' | 'UPDATE';
  error: string;
};

type TodoContextType = {
  todos: Todo[];
  isAdding: boolean;
  addTodo: (todo: Todo) => void;
  deleteTodo: (id: string) => void;
  markAsCompleted: (id: string) => void;
  error: TodoError | null;
  isTodosLoading: boolean;
};

export const TodoContext = createContext<TodoContextType | undefined>(undefined);

export const TodoProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [todos, setTodos] = useState<Todo[]>([]);
  const [isAdding, setIsAdding] = useState(false);
  const [error, setError] = useState<TodoError | null>(null);
  const [isTodosLoading, setIsTodosLoading] = useState(false);

  const addTodo = useCallback(async (todo: Todo) => {
    try {
      setIsAdding(true);
      await createTodo(todo);
      setTodos(currentTodos => [todo, ...currentTodos]);
    } catch {
      setError({
        type: 'ADD',
        error: 'something went wrong creating the todo, please try again later',
      });
    } finally {
      setIsAdding(false);
    }
  }, []);

  const deleteTodo = useCallback(async (id: string) => {
    try {
      await removeTodo(id);
      setTodos(currentTodos => currentTodos.filter(todo => todo.id !== id));
    } catch {
      setError({
        type: 'DELETE',
        error: "sorry, we can't perform this actions right now, please try again later",
      });
    }
  }, []);

  const markAsCompleted = useCallback(async (id: string) => {
    try {
      await markTodoAsComplete(id);
      setTodos(currentTodos =>
        currentTodos.map(todo => (todo.id === id ? { ...todo, completed: true } : todo))
      );
    } catch {
      setError({
        type: 'UPDATE',
        error: "sorry, we can't perform this actions right now, please try again later",
      });
    }
  }, []);

  const loadTodos = useCallback(async () => {
    try {
      setIsTodosLoading(true);
      const data = await getTodos();
      setTodos(data);
    } catch {
      setError({
        type: 'LOAD',
        error: "sorry, we can't fetch your todos right now, please try again later",
      });
    } finally {
      setIsTodosLoading(false);
    }
  }, []);

  useEffect(() => {
    loadTodos();
  }, [loadTodos]);

  return (
    <TodoContext.Provider
      value={{ todos, addTodo, markAsCompleted, deleteTodo, error, isAdding, isTodosLoading }}
    >
      {children}
    </TodoContext.Provider>
  );
};
