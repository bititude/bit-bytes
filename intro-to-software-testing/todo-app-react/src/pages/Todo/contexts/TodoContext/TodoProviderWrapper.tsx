import { TodoProvider } from './TodoContext';

export const TodoProviderWrapper: React.FC = ({ children }) => (
  <TodoProvider>{children}</TodoProvider>
);
