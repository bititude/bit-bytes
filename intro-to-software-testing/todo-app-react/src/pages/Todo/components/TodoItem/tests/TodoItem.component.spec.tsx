/**
 * Requirements
 * 1. pending todos should contain a mark as completed button and delete button
 * 2. completed todos should be striked
 * 3. completed todos should only contain a delete button
 */

import { describe, expect, it } from 'vitest';
import TodoItem from '../TodoItem';
import { TodoProviderWrapper } from '../../../contexts/TodoContext/TodoProviderWrapper';
import { Todo } from '../../../../../types/Todo';
import { faker } from '@faker-js/faker';
import { render, screen } from '@testing-library/react';

describe('TodoItem.tsx', () => {
  it('pending todos should contain a mark as completed button and delete button', () => {
    const pendingTodo: Todo = {
      id: faker.string.uuid(),
      completed: false,
      title: faker.lorem.lines(1),
    };

    render(<TodoItem todo={pendingTodo} />, { wrapper: TodoProviderWrapper });

    const markAsCompletedButton = screen.getByTestId('todo-markascomplete-btn');

    const deleteTodoButton = screen.getByTestId('todo-delete-btn');

    expect(markAsCompletedButton).toBeInTheDocument();

    expect(deleteTodoButton).toBeInTheDocument();
  });

  it('completed todos should be striked', () => {
    const completedTodo: Todo = {
      id: faker.string.uuid(),
      completed: true,
      title: faker.lorem.lines(1),
    };

    render(<TodoItem todo={completedTodo} />, { wrapper: TodoProviderWrapper });

    const strikedTodo = screen.getByTestId('striked-todo');

    expect(strikedTodo).toBeInTheDocument();
  });

  it('completed todos should only contain a delete button', () => {
    const completedTodo: Todo = {
      id: faker.string.uuid(),
      completed: true,
      title: faker.lorem.lines(1),
    };

    render(<TodoItem todo={completedTodo} />, { wrapper: TodoProviderWrapper });

    const markAsCompletedButton = screen.queryByTestId('todo-markascomplete-btn');

    const deleteTodoButton = screen.getByTestId('todo-delete-btn');

    expect(markAsCompletedButton).toBe(null);

    expect(deleteTodoButton).toBeInTheDocument();
  });

  describe('should match snapshot', () => {
    it('should match pending todo snapshot', () => {
      const pendingTodo: Todo = {
        id: faker.string.uuid(),
        completed: false,
        title: 'brush teeth',
      };

      const { asFragment } = render(<TodoItem todo={pendingTodo} />, {
        wrapper: TodoProviderWrapper,
      });

      expect(asFragment()).toMatchSnapshot();
    });

    it('should match completed todo snapshot', () => {
      const completedTodo: Todo = {
        id: faker.string.uuid(),
        completed: true,
        title: 'buy milk',
      };

      const { asFragment } = render(<TodoItem todo={completedTodo} />, {
        wrapper: TodoProviderWrapper,
      });

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
