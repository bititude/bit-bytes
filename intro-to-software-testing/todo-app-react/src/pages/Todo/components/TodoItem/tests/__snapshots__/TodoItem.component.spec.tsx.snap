// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoItem.tsx > should match snapshot > should match completed todo snapshot 1`] = `
<DocumentFragment>
  <div
    class="border-2 border-gray-400 py-2 px-4 font-bold rounded-md shadow-md mb-4 flex items-center justify-between w-full"
    data-testid="todo"
  >
    <p
      class="w-4/5 text-gray-500"
      data-testid="striked-todo"
    >
       
      <s
        class="decoration-2"
      >
        <i>
          buy milk
        </i>
      </s>
    </p>
    <button
      class="rounded-full bg-red-500   w-6 h-6 flex items-start justify-center hover:scale-105 "
      data-testid="todo-delete-btn"
    >
      <span
        class=" text-white font-extrabold "
      >
        ×
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`TodoItem.tsx > should match snapshot > should match pending todo snapshot 1`] = `
<DocumentFragment>
  <div
    class="border-2 border-gray-400 py-2 px-4 font-bold rounded-md shadow-md mb-4 flex items-center justify-between w-full"
    data-testid="todo"
  >
    <p
      class="w-4/5"
    >
      brush teeth
    </p>
    <button
      class="rounded-full bg-red-500   w-6 h-6 flex items-start justify-center hover:scale-105 "
      data-testid="todo-delete-btn"
    >
      <span
        class=" text-white font-extrabold "
      >
        ×
      </span>
    </button>
    <button
      class="rounded-full bg-green-500 ml-3  w-6 h-6 flex items-start justify-center hover:scale-105 "
      data-testid="todo-markascomplete-btn"
    >
      <span
        class=" text-white font-extrabold "
      >
        ✓
      </span>
    </button>
  </div>
</DocumentFragment>
`;
